pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -121,325 +121,72 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
-	// object failed to create, requeue
-	if err != nil {
-		return result, err
-	}
-
-	// Define a new KWebUI object
-	kwebuiCR := newKWebUIForCR(instance)
-
-	// Set HyperConverged instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, kwebuiCR, r.scheme); err != nil {
-		return reconcile.Result{}, err
-	}
-
-	// Create the KWebUI CR if it doesn't already exist
-	result, err = manageComponentResource(kwebuiCR, "KWebUI", r.client)
-
-	// KWebUI failed to create, requeue
-	if err != nil {
-		return result, err
-	}
-
-	// Everything went fine, automatically reconcile after after a minute without observed activity to
-	// make sure that even deployed objects without owner reference will be re-created if removed.
-	// TODO djzager: What I think we should do is to lock down the HCO CR to a specified name
-	// (via environment variable on the operator deployment) and a specified namespace (can use the
-	// downward API to set an environment variable on the operator deployment getting the namespace
-	// where the operator was deployed).
-	return reconcile.Result{RequeueAfter: time.Minute}, nil
+	return r.reconcileUpdate(reqLogger, instance, request)
 }
 
-func manageComponentResource(o runtime.Object, kind string, c client.Client) (reconcile.Result, error) {
-	err := c.Create(context.TODO(), o)
-	if err != nil && errors.IsAlreadyExists(err) {
-		log.Info("Skip reconcile: resource already exists", "Kind", kind)
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
-	}
-
-	log.Info("Creating new resource", "Kind", kind)
-	return reconcile.Result{}, nil
+// ResultFromCRCreate holds the result and error from creating a CR
+type ResultFromCRCreate struct {
+	result reconcile.Result
+	err    error
 }
 
-func manageComponentResourceRemoval(o interface{}, c client.Client, cr *hcov1alpha1.HyperConverged) (reconcile.Result, error) {
-	resource, err := toUnstructured(o)
-	if err != nil {
-		log.Error(err, "Failed to convert object to Unstructured")
-		return reconcile.Result{}, err
-	}
-
-	err = c.Get(context.TODO(), types.NamespacedName{Name: resource.GetName(), Namespace: resource.GetNamespace()}, resource)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			log.Info("Resource doesn't exist, there is nothing to remove", "Kind", resource.GetObjectKind())
-			return reconcile.Result{}, nil
-		}
-		return reconcile.Result{}, err
-	}
+func (r *ReconcileHyperConverged) reconcileUpdate(logger logr.Logger, cr *hcov1alpha1.HyperConverged, request reconcile.Request) (reconcile.Result, error) {
+	results := []ResultFromCRCreate{}
 
-	labels := resource.GetLabels()
-	if app, labelExists := labels["app"]; !labelExists || app != cr.Name {
-		log.Info("Existing resource wasn't deployed by HCO, ignoring", "Kind", resource.GetObjectKind())
-		return reconcile.Result{}, nil
-	}
+	resources := r.getAllResources(cr, request)
```

Comment:
To be honest, I find [this portion of the knicluster_controller reconcile](https://github.com/mhrivnak/kni-operator/blob/master/pkg/controller/knicluster/knicluster_controller.go#L272-L281) to be similar to what you are trying to accomplish here while being somewhat easier to follow and reason about.

Maybe we can update the `newKubevirtTemplateValidatorForCR` type functions to instead ensure that a given resource is as it should be. Then in the reconcile we just loop through those functions.

Reply:
Another advantage of this approach is that if it gets more complex than "create if it doesn't exist" then that complexity is in the function where that logic belongs.

Reply:
@djzager I submitted an update. It should look more like the knicluster_controller now.